crates/cext/src/sparse_observable.rs

```diff
@@ -820,6 +821,58 @@ pub unsafe extern "C" fn qk_obs_equal(
     obs.eq(other)
 }
 
+/// @ingroup QkObs
+/// Apply a [NLayout] to the observable.
+///
+/// @param obs A pointer to the observable.
+/// @param layout A pointer to the layout.
+///
+/// @return The observable with the layout applied, or null if the layout was incompatible
+///     with the observable.
+///
+/// # Example
+///
+///     QkObs *obs = qk_obs_zero(4);
+///
+///     QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+///     uint32_t qubits[3] = {1, 2, 3};
+///     complex double coeff = 1;
+///     QkObsTerm term = {coeff, 3, bit_terms, qubits, 4};
+///
+///     int err = qk_obs_add_term(obs, &term);
+///     if (err != 0) {
+///         qk_obs_free(obs);
+///         return err;
+///     }
+///
+///     uint32_t virt_to_phys[3] = {0, 3, 2, 1};
+///     QkLayout *layout = qk_layout_new(virt_to_phys, num_qubits);
+///     QkObs *obs_with_layout = qk_obs_apply_layout(obs, layout);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` is not a valid, non-null pointer to ``QkObs`` or if ``layout``
+/// is not a valid, non-null pointer to ``QkLayout``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_apply_layout(
+    obs: *const SparseObservable,
+    layout: *const NLayout,
+) -> *mut SparseObservable {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let layout = unsafe { const_ptr_as_ref(layout) };
+
+    // Cast the slice of PhysicalQubit (which is u32) to plain u32, which is required
+    // by the apply_layout method.
+    let as_slice: &[u32] = ::bytemuck::cast_slice(&layout.virt_to_phys);
```

Comment:
This line (and the `NoUninit` for `PhysicalQubit`) warrants extra care since it's the first time I'm using `bytemuck` 😄